File: c/linear_hash.c
/**
 * LINEAR HASH
 *   線形二乗アルゴリズムを使用
 *   Rubyと違ってC言語だと mod に比べてかなり早くなる
 *
 * linear hash: 3.608443
 * hash(mod): 5.345495
 */

#ifdef WIN32

#include <windows.h>
double get_time()
{
    LARGE_INTEGER t, f;
    QueryPerformanceCounter(&t);
    QueryPerformanceFrequency(&f);
    return (double)t.QuadPart/(double)f.QuadPart;
}

#else

#include <sys/time.h>
#include <sys/resource.h>

double get_time()
{
    struct timeval t;
    struct timezone tzp;
    gettimeofday(&t, &tzp);
    return t.tv_sec + t.tv_usec * 1e-6;
}

#endif

#include <stdio.h>
#include <math.h>
#define TIMES 2000000000

void linear_hash(int num)
{
	int n;
	int result;

	for(n = 0; n < TIMES; n++)
	{
		/* linear hash */
		result = n & (num - 1);
	}
}

void hash(int num)
{
	int n;
	int result;

	for(n = 0; n < TIMES; n++)
	{
		/* hash(mod) */
		result = n % num;
	}
}

int main()
{
	int n;
	int result;
	double begin1;
	double end1;
	double begin2;
	double end2;
	const int num = (int)pow(2, 2);

	begin1 = get_time();
	linear_hash(num);
	end1 = get_time();
	printf("linear hash: %f\n", end1 - begin1);

	begin2 = get_time();
	hash(num);
	end2 = get_time();
	printf("hash(mod): %f\n", end2 - begin2);

	return 0;
}

File: ruby/linear_hash.rb
# LINEAR HASH
#   線形二乗アルゴリズムを使用
#   僅かにLINEAR HASHの方が早いが誤差の範疇のような感じ
#   分割数が2のべき乗でないとLINEAR HASHの方が遅い
#   C言語とかならもっと違いが出るかも
require 'benchmark'

def least_squares(num)
  return 0 if num == 0
  return 1 if num == 1
  2 ** Math.log(num, 2).ceil
end

def linear_hash(n, v, num)
  n & (v - 1)
#  return t if t < num
#  t & ((v / 2).ceil - 1)
end

def hash(n, num)
  n % num
end

TIMES = 1000000000
DIVISION_NUMBER = 2 ** 6
Benchmark.bm 11 do |r|
  r.report "LINEAR HASH" do
    v = least_squares(DIVISION_NUMBER)
    TIMES.times do |n|
      linear_hash(n, v, DIVISION_NUMBER)
    end
  end

  r.report "HASH" do
    TIMES.times do |n|
      hash(n, DIVISION_NUMBER)
    end
  end
end
#                   user     system      total        real
# LINEAR HASH  43.945000   0.000000  43.945000 ( 43.986806)
# HASH         44.179000   0.000000  44.179000 ( 44.259180)
